### PR TITLE
Machanism for filter annotation&label passed from Application to workload and trait

### LINKED ISSUE #1801: mechanism for filter annotations or labels that should not be passed through workloads/traits
**Is your feature request related to a problem? Please describe.**

Now annotations and labels from Application（and AppConfig） will be passed through workloads and traits.

But these annotations and labels are not all necessary and some of them may contain sensitive information that should be filtered.

**Describe the solution you'd like**

We should provide an mechanism to filter. I suggest we should add special annotations, which can contain filter keys.

```
kind: Application
metadata:
  annotations:
     filter.oam.dev/annotation-keys: "specialKey1,specialKey2"
     filter.oam.dev/label-keys: "specialKey1,specialKey2"
```

```
kind: ApplicationConfiguration
metadata:
  annotations:
     filter.oam.dev/annotation-keys: "specialKey1,specialKey2"
     filter.oam.dev/label-keys: "specialKey1,specialKey2"
```

When these keys are specified, the labels and annotations should be filtered by these keys.

Please separate these into two PRs, one for Application, and one for ApplicationConfiguration.

### LINKED ISSUE #1836: [Feature] refactor vela cli env command to use Initializer
**Is your feature request related to a problem? Please describe.**

Currently the vela cli just use local files, and create resources such as namespace directly by client.

And now we have [iniitializer](https://github.com/oam-dev/kubevela/blob/master/design/vela-core/environment.md) design and [implementation](https://github.com/oam-dev/kubevela/pull/1829).

We should refactor vela cli to use the initializer.

**Describe the solution you'd like**

1. don't create local files. Create resources and keep metadata in initializer.
2. List/switch environments by get initializer objects from cluster.
3. Get namespace from initializer object.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by chivalryq**
**What this PR does / why we need it**:

ref to #1801, add two annotations to specify annotations and labels which shouldn't be passed to workload&trait. 

Two spectial annotation are: `filter.oam.dev/annotation-keys` and `filter.oam.dev/label-keys`, they specify the annotations and label keys shouldn't be passed to workloads and traits. Multiple keys can be split by comma.

Also, the two special annotations themselves won't be passed to workload and trait.

Now you can apply an application yaml like below:

```yaml
apiVersion: core.oam.dev/v1beta1 
kind: Application 
metadata: 
  name: filter-demo
  annotations: 
    filter.oam.dev/annotation-keys: "notPassAnno1, notPassAnno2" 
    filter.oam.dev/label-keys: "notPassLabel" 
    notPassAnno1: "this-annotation-can-not-be-passed" 
    notPassAnno2: "this-annotation-can-not-be-passed-too" 
    canPassAnno: "Annotation-passed" 
  labels: 
    notPassLabel: "Label-filtered" 
    canPassLabel: "Label-passed" 
spec: 
  ...
```

**How this PR do**

Given that we don't use app-config anymore, I just put it in assemble phase

> applyAppManifetsts
> * if it's revision-only, then just createOrUpdate ResourcesConfigMap
> * if it's rollout-template, then just return
> * dispatch workload addon resources
>   
>   * workload addon means HelmRelease/HelmRepo of Helm-based component. we use dispatch instead of client.Apply to emit these resources now.
> * assemble & dispatch app manifests  **<--added here**
>   
>   * do GC after dispatching

ref [render phase](https://github.com/oam-dev/kubevela/pull/1814#issuecomment-863973759)

**Which issue(s) this PR fixes**:

Fixes #1801

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by captainroy-hy** [comment]
I think it's related to #1801 but not #1836 ?

File: pkg/oam/util/helper.go [hunk 1]
```diff
@@ -475,10 +475,39 @@ func PassLabel(parentObj, childObj labelAnnotationObject) {
 // PassLabelAndAnnotation passes through labels and annotation objectMeta from the parent to the child object
 // when annotation or labels has conflicts, the parentObj will override the childObj.
 func PassLabelAndAnnotation(parentObj, childObj labelAnnotationObject) {
```
Comment (captainroy-hy) [inline pkg/oam/util/helper.go:477]: This util function should not have business-specific logic such as filter labels&annots of app.
I would like to suggest you implement this logic [here](https://github.com/oam-dev/kubevela/blob/b582ccc0600dc5cb6c313015ac7431cef3e057e5/pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go#L203-L205). This is where collecting app's lables&annots being passed to workload/trait. Filter lables&annots in that method.

Comment (captainroy-hy) [reply pkg/oam/util/helper.go:477]: Additionally, I made a simple annotation filter [here](https://github.com/oam-dev/kubevela/blob/b582ccc0600dc5cb6c313015ac7431cef3e057e5/pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go#L256). If necessary, you can remove it and merge this logic into yours.

Comment (captainroy-hy) [reply pkg/oam/util/helper.go:477]: The other option is [here ](https://github.com/oam-dev/kubevela/blob/b582ccc0600dc5cb6c313015ac7431cef3e057e5/pkg/controller/core.oam.dev/v1alpha2/application/revision.go#L436-L438)where passing lables&annots to appRevision. This a path how label&annots propagated:  app => appRevision => workload/trait.  So if you need filter these label&annots from appRevision, do it here. It's up to your requirement.

File: pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go [hunk 2]
```diff
@@ -236,28 +238,63 @@ func (am *AppManifests) validate() error {
...
+	}
+	filter := func(before map[string]string, notAllowedKey []string) map[string]string {
+		after := make(map[string]string)
+		for key, value := range before {
+			if !in(key, notAllowedKey) {
+				after[key] = value
+			}
+		}
+		return after
+	}
```
Comment (captainroy-hy) [inline pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go:258]: Why not just use a loop inside which invoke `delete(labels, filteredLabelKey)`

Comment (captainroy-hy) [reply pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go:258]: For each filterLabelKey, it's better to trim spaces before filtering. I think it's highly possible to type spaces between keys in the annotation.

File: pkg/oam/labels.go [hunk 3]
```diff
@@ -99,4 +99,10 @@ const (
 
 	// AnnotationKubeVelaVersion is used to record current KubeVela version
 	AnnotationKubeVelaVersion = "oam.dev/kubevela-version"
+
+	// AnnotationFilterAnnotationKeys is used to filter annotations passed to workload and trait
```
Comment (captainroy-hy) [inline pkg/oam/labels.go:103]: Indicate that keys are joint by `,` in the annotation.

**Review by captainroy-hy — CHANGES_REQUESTED** [review]

File: pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go (hunk 2 above)
Comment (chivalryq) [reply pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go:258]: yeah, that's better

File: pkg/oam/labels.go (hunk 3 above)
Comment (chivalryq) [reply pkg/oam/labels.go]: done

File: pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go [hunk 4]
```diff
@@ -236,28 +238,52 @@ func (am *AppManifests) validate() error {
 }
 
 // workload and trait in the same component both have these labels
-func (am *AppManifests) generateCommonLabels(compName, compRevisionName string) map[string]string {
+func (am *AppManifests) generateAndFilterCommonLabels(compName, compRevisionName string) map[string]string {
+	filter := func(labels map[string]string, notAllowedKey []string) {
+		for _, _l := range notAllowedKey {
+			l := strings.Trim(_l, " ")
+			delete(labels, l)
```
Comment (captainroy-hy) [inline pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go:245]:
```suggestion
			delete(labels, strings.TrimSpace(_l))
```

**Comment by captainroy-hy** [comment]
LGTM after adding unit tests and passing CI.

File: pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go (hunk 4 above)
Comment (chivalryq) [reply pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go]: done

**Review by captainroy-hy — APPROVED** [review]

**Comment by captainroy-hy** [comment]
LGTM

File: pkg/controller/core.oam.dev/v1alpha2/application/assemble/testdata/filter_annotations.yaml [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+                spec:
+                  containers:
+                    - image: nginx
+                      name: frontend
+                      ports:
+                        - containerPort: 80
+        status:
+          observedGeneration: 0
+  resourcesConfigMap:
+    name: website-v1
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/assemble/testdata/filter_annotations.yaml:189]: please clean unrelated fields in this test data

File: pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go [hunk 6]
```diff
@@ -236,28 +238,51 @@ func (am *AppManifests) validate() error {
...
+	finalLabels := util.MergeMapOverrideWithDst(Labels, am.appLabels)
+	filterLabels, ok := am.appAnnotations[oam.AnnotationFilterLabelKeys]
+	if ok {
+		filter(finalLabels, strings.Split(filterLabels, ","))
+	}
+	return finalLabels
+}
+
+// DefaultFilterAnnots are annotations won't pass to workload or trait
+var DefaultFilterAnnots = []string{
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go:264]: 一般全局变量放在文件最上面。

File: pkg/controller/core.oam.dev/v1alpha2/application/assemble/testdata/filter_annotations.yaml (hunk 5 above)
Comment (chivalryq) [reply pkg/controller/core.oam.dev/v1alpha2/application/assemble/testdata/filter_annotations.yaml]: done

File: pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go (hunk 6 above)
Comment (chivalryq) [reply pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go]: done

**Review by wonderflow — APPROVED** [review]

**Pull request merged**